Na values

katie-lamb opened the pull request:
The following columns are filled with `pd.NA` in their respective datasets when going through the extract and transform steps of the pipeline. 
### FERC 1

- `construction_type`  
### EIA 

- `natural_gas_delivery_contract_type_code`
- `natural_gas_transport_code`

The call `df.to_sql()` in `sqlite.py` converts the `pd.NA` values to NoneType since the column dtypes are not preserved. This problem is fixed after the fact by recasting the column to a string type once pudl.sqlite is read into a pandas dataframe. Preserving the column dtype is maybe something that should be addressed at some point.

The following columns are filled with `np.nan` in the EPA CEMS datasets, not `pd.NA`. This is because they are categorical type columns and thus do not allow for `pd.NA`. 
### EPA CEMS

- `co2_mass_measurement_code`
- `nox_mass_measurement_code`
- `nox_rate_measurement_code`
- `so2_mass_measurement_code`

### Note
The column `purchase_type` is listed in the issue as an EIA column with an empty string in the enum. However this appears to be a FERC 1 column and the enum is based off the dictionary `POWER_PURCHASE_TYPES_FERC1` in `labels.py`. There is no empty string listed in the `POWER_PURCHASE_TYPES_FERC1` dictionary, so I didn't do anything about the `purchase_type` column.

katie-lamb commented:
Forgot to address this in above comments - I could also turn "unknown" into NAs for these columns if wanted

zaneselvans reviewed (changes requested):
Mostly just some questions about whether there are additional missing-ish values that we should turn into NAs.

zaneselvans commented:
> The call df.to_sql() in sqlite.py converts the pd.NA values to NoneType since the column dtypes are not preserved. This problem is fixed after the fact by recasting the column to a string type once pudl.sqlite is read into a pandas dataframe. Preserving the column dtype is maybe something that should be addressed at some point.

I think what's happening here is that the null values are getting loaded into SQLite just fine, and stored as appropriate NULL values as far as the database is concerned, but then pandas is doing it's standard datatype conversions when we call `pd.read_sql()` and it incorrectly converts the SQLite NULL values into `None` (which is often treated as a "null" object).

With the new metadata system we should be able to give pandas an explicit schema that it can use to type the columns it reads out of the DB appropriately, but that's a later issue.

katie-lamb commented on src/pudl/transform/ferc1.py:
```diff
@@ -769,10 +769,8 @@ def _plants_steam_clean(ferc1_steam_df):
-              [CONSTRUCTION_TYPE_STRINGS, PLANT_KIND_STRINGS],
-              unmapped='')
+        .pipe(pudl.helpers.cleanstrings, ['construction_type'], [CONSTRUCTION_TYPE_STRINGS], unmapped=pd.NA)
+        .pipe(pudl.helpers.cleanstrings, ['plant_type'], [PLANT_KIND_STRINGS], unmapped="")
```
`plant_type` wasn't listed as a column that needed to remove empty strings in the issue, so when I encountered this I skipped `plant_type`. Also a few tests failed when I changed it to pd.NA (I forget what the error is but could regenerate), maybe the column is treated as a primary key? I could go back and change this to NAs though

katie-lamb replied:
In `fields.py` there's this comment: 

TODO Disambiguate column name and apply table specific ENUM constraints. There are different allowable values in different tables.

So it seems like maybe this is a good candidate to remove the empty string?

zaneselvans replied:
Ah okay. I don't think it's ever used as a primary key. There's a mechanism for overriding the generic purely field-name based attributes if we need to. It looks like `plant_type` shows up in in the `plants_hydro_ferc1`, `plants_small_ferc1`, and `plants_steam_ferc1` tables, but currently only the `plants_steam_ferc1` version is being cleaned and set to some canonical values defined in the `PLANT_KIND_STRINGS` , so to do what the TODO asks would require categorizing all the (different) nonsense strings in the hydro and small plants tables. I don't know if this is something that might be getting done already in @aesharpe's FERC 1 cleaning PR but we should ask.

I think the thing to do for now if we want to apply a table specific ENUM is to override the constraints which are defined for this column only for the `plants_steam_ferc1` table, using the `FIELD_METADATA_BY_RESOURCE` defined at the bottom of the `fields.py` module (currently there's only one override). And then maybe create an issue to remind us to create ENUM constraints for the `plant_type` field in the hydro and small plants tables in concert with Austen's cleanup of those tables.

aesharpe replied:
Right now, the `plant_type` column gets cleaned and standardized in the steam table, manually mapped in the small generators table, left alone in the hydro table, and is non-existent (and irrelevant) in the pumped storage table.

Part of my FERC 1 cleaning module extracts the headers from the small generators table, standardizes them, and puts them in the `plant_type` column so that we don't have to rely on manual mapping. This is still a draft PR, however, so none of this is integrated yet. 

The `plant_type` extracted from the headers is standardized according to a dictionary that I created and which we can change to reflect whatever final plant type string we want (or specify in an ENUM constraint). 

Current small plant fuel types: 
```
None, 'hydro', 'internal_combustion', 'solar_pv', 'cogen',
       'steam_heat', 'wind', 'gas_turbine', 'diesel_turbine',
       'pumped_hydro', 'coal', 'waste_heat', 'fuel_cell'
```

Current steam fuel types: 
```
'steam', 'nuclear', 'unknown', 'combustion_turbine',
       'combined_cycle', 'internal_combustion', 'geothermal', 'wind',
       'photovoltaic', 'solar_thermal', ''
```

At the very least, we could standardize the NA values for both of these tables. Ideally we will pull values for `plant_type` from the same set of values, whether they are constrained or not. I think it's a good idea to create an issue to remind us to do this (and for the hydro table and maybe even an arbitrary one for the pumped storage table so that it has a `plant_type` when it gets merged with the others into the mega FERC table). Ultimately, the goal of all this RMI manual mapping we're doing is to connect, among other things, the EIA `technology_description` field with individual FERC records so we won't need to use the janky FERC `plant_type` or header extraction at all.

katie-lamb replied:
For now I've moved `plant_type` to the FIELD_METADATA_BY_RESOURCE and added an assert so that there are no uncaught strings in the `plant_type` column of `plants_steam_ferc1`. Seems like there should be a new issue that adds ENUM constraints for the FERC 1 hydro and small plants tables? Or is the goal to remove `plant_type` altogether and have it agree with EIA `technology_description`?

katie-lamb commented:
I made a list of the fields with `unknown` or `undetermined` as well as some of the strings that map to these values: 

## FERC1
### construction_type 
'', 'automatic operation', 'comb. turb. installn', 'comb. turb. instaln',
        'com. turb. installn', 'n/a', 'for detailed info.', 'for detailed info',

### plant_kind
  
'', 'n/a', 'see pgs 402.1-402.3', 'see pgs 403.1-403.9', "respondent's share",
        '--', '(see note 7)', 'other', 'not applicable', 'peach bottom', 'none.',
        'fuel facilities', '0', 'not in service', 'none', 'common expenses', ..... many more

### fuel_type

'', '1265', 'mwh units', 'composite', 'therms', 'n/a', 'mbtu/kg', 'uranium 235',
'oil', 'ccf', '2261', 'uo2', '(7)', 'oil #2', 'oil #6', '\x99å\x83\x90?"', .... many more

- `construction_type` doesn't seem like a primary key and maybe is a good candidate to make NAs? 
## EPA CEMS
`co2_mass_measurement_code`
`nox_mass_measurement_code`
`nox_rate_measurement_code`
`so2_mass_measurement_code`  
All accept "other", "undetermined", or "unknown code" - these could probably be made into NAs?

zaneselvans commented:
## FERC 1
The FERC 1 `construction_type`, `plant_kind` and `fuel_unit` "cleaned strings" are things we assigned, so we won't mess up anyone else if we make them NA, but we might mess up our own code. There's an `unmapped` argument in `pudl.helpers.cleanstrings()` that we could use to set any unmapped strings to `pd.NA` but I don't think we want to do that, because we want it to be clear whether we've reviewed and attempted to categorize every one of these nonsense strings, and they add new ones every year. So we may want to go ahead and assign them `unknown` assert that every value in the column is one of our simplified values (so we catch anything new and unexpected) and then replace `unknown` with `pd.NA`.

## EIA
We should look at what kinds of utilities are being assigned "unknown" and "other" entity types, and whether they are consistent across years, or if how they're reported has changed. Like did they used to be "unknown" and now they're "other"? Or did they eventually get categorized? What proportion of the overall population has these mystery types? how often does a utility entity type change? Can we fill in the missing values confidently? What does the EIA documentation say about this field (if anything)? Often this is in one of the non-data tabs of the raw input spreadsheets, or in the instructions (PDFs which we've tried to archive in our docs)

## EPA CEMS
Similarly we should see what EPA is saying these fields / values mean, and try to understand if "undetermined" and "unknown code" and "other" mean different things, or if they've just changed the coding scheme over the years, and also what proportion of the dataset is affected. There's probably some documentation on web from EPA, but I'm not 100% sure where.  If we can't find it, we've been in contact with some of the EPA people in the past and might be able to ask for help.

katie-lamb commented on src/pudl/metadata/enums.py:
```diff
@@ -214,6 +214,5 @@
     "Substitute",
     "Undetermined",  # Should be replaced with NA
     "Unknown Code",  # Should be replaced with NA
```
This is the EPA CEMS measurement codes - I couldn't really find documentation about this but it seems like the "Undetermined" and "Unknown Code" are fields created by EPA. I didn't check all the years and states but from a small pass through the data I didn't find any data with these codes. Ultimately, I think we should just leave them for now.

katie-lamb commented on src/pudl/transform/eia860.py:
```diff
@@ -579,6 +579,7 @@ def _make_phone_number(col1, col2, col3):
             entity_type=lambda x: x.entity_type.map(ENTITY_TYPES)
         )
         .pipe(pudl.helpers.convert_to_date)
+        .fillna({'entity_type': pd.NA})
```
This changes np.nans to pd.NAs. There are no "unknown" in the EIA 860 entity types for all years.

zaneselvans replied:
I know this field comes up a lot in the EIA-861, so I wonder if this is where the `unknown` and `other` values are showing up? I'm working on compiling that metadata right now, and for some reason 100% of the `entity_type` column in the `sales_eia861` table is currently showing up as null.

But if this all works for the existing EIA related DB tables then that's fine.

zaneselvans replied:
I'm going to have to do some messing around with `entity_type` in the EIA-861 (which is where it mostly shows up) and might end up cleaning up the categories. It looks like they use two different sets of codes in there. The short codes and full names with capitalization, which is why I'm getting fully null columns.

katie-lamb commented on src/pudl/transform/ferc1.py:
```diff
@@ -1075,6 +1087,8 @@ def fuel(ferc1_raw_dfs, ferc1_transformed_dfs):
     fuel_ferc1_df.dropna(inplace=True)
 
+    # Replace "unkown" fuel unit with NAs - this comes after we drop missing data with NAs
+    fuel_ferc1_df = fuel_ferc1_df.replace({'fuel_units': 'unknown'}, pd.NA)
```
The line above drops all the NAs from the dataframe. I assume we want to keep data that has "unknown" as the fuel type so I replaced the "unknown" fuel types with `pd.NA`s after the `dropna` call. If we want to drop "unknown" fuel types as well then I can just move this `replace` call above `dropna`

zaneselvans replied:
Wow, that blanket `dropna()` is kind of scary, and probably leftover from very early in the project. I wonder if this is something that @aesharpe has changed in her recent work on the FERC 1 cleaning? This probably requires a closer look at what all is being kept / dropped in the `fuel_ferc1` table during its transformation to see what will be reasonable -- i.e. what criteria do we want to apply for retaining a fuel record -- requiring that it have no NA values seems like it might be excluding too many records.

katie-lamb replied:
I saw someone (probably Austen) left a new comment above this `dropna()` call but seems like it's staying for the mean time

zaneselvans replied:
Yeah, it's a different issue really. We should make an issue and re-visit it in other FERC cleaning like you mentioned in your slack message.

katie-lamb commented:
Summary of some changes: 

### FERC 1

I find some strings in `construction_type`, `plant_type`, and `fuel_unit` that hadn't been caught yet by the dictionary or strings and put them in the `unknown` category of this dictionary. I also added asserts so that we know if there are more uncaught strings. Then I converted all the `unknown` to `pd.NA` after this assert is made. 

### EIA 860

`entity_type` has `unknown` as a valid value, however there aren't actually any data points that use this `unknown` field. I couldn't find much in the documentation about `entity_type` but `other` seems to be an EIA label, so I left all the `other` values in the dataframes. There are some data points (58,000 out of around 109,000 rows for all years) that have `np.nan` so I converted these to `pd.NA`

### EPA CEMS

I couldn't find much documentation about the nox and co2 measurement codes, but it seems like the `Undetermined` and `Unknown Code` are rarely used, so I left them.

zaneselvans reviewed (commented):
Made a few comments, if you want to chat. I suspect we should look at the broad dropna() from the `fuel_ferc1` table more closely.

katie-lamb commented on src/pudl/metadata/fields.py:
```diff
@@ -2324,5 +2323,13 @@
+        "plant_type": {
+            "type": "string",
+            "constraints": {
+                "enum": ['steam', 'combustion_turbine', 'combined_cycle', 'nuclear', 'geothermal', 'internal_combustion', 'wind', 'photovoltaic', 'solar_thermal']
```
Changed this to use the keys of the string cleaning dictionaries for all the cleaned up categorical columns

zaneselvans replied:
Ah, that's annoying. I feel like I *still* don't really understand how Python imports are supposed to work and get bitten by this.
